### DISCUSSION TITLE
"mogrify -rotate" webp from within PHP = Error mogrify: delegate failed 'dwebp'

### DISCUSSION THREAD

**Discussion by futureweb** [ImageMagick/ImageMagick#4239] [Help]
Hey there,
I got a little problem with using mogrify -rotate from within PHP.

When called from Linux CLI directly it's working:
```
# /usr/bin/mogrify -rotate "90" '1.webp' 
Decoded /tmp/magick-147387ncd0y7XkoPex. Dimensions: 368 x 550 . Format: lossy. Now saving...
Saved file /tmp/magick-147387ucsuNo6blGDG
```

But when I try to call it from PHP using exec() or system() it's failing with:

$cmd = '/usr/bin/mogrify -verbose -rotate "' . $rotate . '" ' . escapeshellarg($filepath).' 2>&1';
exec($cmd, $orientateReturn, $retval);

```
 'dwebp' -pam '/tmp/magick-1465581Xug1-4XX4nZ' -o '/tmp/magick-14655849E7XWU76ETS' mogrify: delegate failed `'dwebp' -pam '%i' -o '%o'' @ error/delegate.c/InvokeDelegate/1928. mogrify: unable to open file `/tmp/magick-146558oKQpevnIrSob': No such file or directory @ error/constitute.c/ReadImage/604. 
```

seems like it's failing as dwebp isn't in the path of the PHP Shell ... can I somehow tell mogrify that it should call dwebp with it's full path?!? Can I tell mogrify the dwebp path somehow? 

thx! ;-)

Andy

**Comment by futureweb** [comment]
ok it doesn't just seem to be a problem with the path. I now went directly to "/ usr / local / bin /" where 'dwebp' is in the PHP script as a test. Then the error changes:

$cmd = 'cd /usr/local/bin/;/usr/bin/mogrify -verbose -rotate "' . $rotate . '" ' . escapeshellarg($filepath).' 2>&1';
exec($cmd, $orientateReturn, $retval);

```
[0] => 'dwebp' -pam '/tmp/magick-151229VHbnVZKcTKqR' -o '/tmp/magick-151229eGl4KOmGPWbg'
    [1] => Decoded /tmp/magick-151229VHbnVZKcTKqR. Dimensions: 550 x 368 . Format: lossy. Now saving...
    [2] => Saved file /tmp/magick-151229eGl4KOmGPWbg
    [3] => /daten/www/xxx/userdata/100/pid/50/yy/88/1_1.webp PAM 550x368 550x368+0+0 8-bit TrueColor sRGB 809669B 0.000u 0:00.004
    [4] => /daten/www/xxx/userdata/100/pid/50/yy/88/1_1.webp=>/tmp/magick-151229EZPQZ_n9b9dJ PNG 550x368=>368x550 368x550+0+0 8-bit sRGB 324674B 0.220u 0:00.185
    [5] => 'cwebp' -quiet 92 '/tmp/magick-1512291cA2NNnyxrGX' -o '/tmp/magick-151229XuAHX7rxNi2k'
    [6] => Error! Cannot open output file '/tmp/magick-151229XuAHX7rxNi2k'
    [7] => /daten/www/xxx/userdata/100/pid/50/yy/88/1_1.webp PNG 550x368=>368x550 368x550+0+0 8-bit sRGB 324674B 0.000u 0:00.025
    [8] => mogrify: delegate failed `'cwebp' -quiet %Q '%i' -o '%o'' @ error/delegate.c/InvokeDelegate/1928.

```

It somehow looks as also the communication between dwebp and mogrify is somehow broken?
Hope someone can give me a helping hand on this! ;-)

thx

**Comment by snibgo** [comment]
It seems your version of IM does not have webp as a built-in (internal) delegate, so it converts to and from webp by calling external delegates (executables).

> Can I tell mogrify the dwebp path somehow?

I don't use PHP, but I think it doesn't know about the system path. The fix is t edit your delegates.xml to give the full path to the dwebp program.


>     [5] => 'cwebp' -quiet 92 '/tmp/magick-1512291cA2NNnyxrGX' -o '/tmp/magick-151229XuAHX7rxNi2k'

That line looks dodgy. What is "92"? Please show the dwebp and cwebp lines from delegates.xml.

IM can be built with webp as an _internal_ delegate. This may be a better solution for you.

**Reply by futureweb** [reply]
it's the RHEL/Centos 7 default ImageMagick i'm using:

```
# mogrify --version
Version: ImageMagick 6.9.10-68 Q16 x86_64 2021-02-24 https://imagemagick.org
Copyright: © 1999-2019 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP(3.1) 
Delegates (built-in): bzlib cairo fontconfig freetype gslib jng jp2 jpeg lcms ltdl lzma openexr pangocairo png ps rsvg tiff wmf x xml zlib

# rpm -qa | grep -i imagemagick
ImageMagick-6.9.10.68-5.el7_9.x86_64
ImageMagick-devel-6.9.10.68-5.el7_9.x86_64
```

will have a look if the Version from REMI Repo got internal webp support - should solve this problem if it got internal webp support ... mhm

> That line looks dodgy. What is "92"? Please show the dwebp and cwebp lines from delegates.xml.

` <delegate decode="png" encode="webp" command="&quot;cwebp&quot; -quiet %Q &quot;%i&quot; -o &quot;%o&quot;"/>`

92 should be the Quality?

**Reply by futureweb** [reply]
btw) I just tried to add the full path to the binary to delegates.xml and it workes ... thank you! :-)

```
<!--  <delegate decode="webp" command="&quot;dwebp&quot; -pam &quot;%i&quot; -o &quot;%o&quot;"/>-->
<delegate decode="webp" command="&quot;/usr/local/bin/dwebp&quot; -pam &quot;%i&quot; -o &quot;%o&quot;"/>
<!--  <delegate decode="png" encode="webp" command="&quot;cwebp&quot; -quiet %Q &quot;%i&quot; -o &quot;%o&quot;"/>-->
<delegate decode="png" encode="webp" command="&quot;/usr/local/bin/cwebp&quot; -quiet %Q &quot;%i&quot; -o &quot;%o&quot;"/>
```

Even I don't fully understand why it's not working without full path to binary in the delegates.xml - in my understanding it should work ask "/usr/local/bin/" is within $PATH in my PHP Shell ... 
```
echo $PATH;
/usr/local/bin:/usr/bin
```

tomorrow I will try to find a Version for Centos 7 with integrated webp support as you suggested ... :-)

**Comment by snibgo** [comment]
I suspect the "%Q" in delegates.xml is a mistake, but harmless. "cwebp --help" shows no parameter in that place. Experiment shows that superfluous words in that position are ignored. For example:
```
cwebp abc trtrtret toes.png -o x.webp
```
This creates x.webp from toes.png, and ignores "abc" and "trtrtret".

**Reply by futureweb** [reply]
interesting ... that is the delegates.xml supplied by the RHEL / CENTOS package ... but that's right ... Quality should be specified with a "-q <float>" ...
